File: src-tauri/src/models/modem.rs
//! Modem wire models: modem, SIM and bearer information.

use serde::{Deserialize, Serialize};

/// A cellular modem as reported to the frontend.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ModemInfo {
    pub id: String,
    pub manufacturer: String,
    pub model: String,
    pub device: String,
    pub imei: Option<String>,
    pub operator: Option<String>,
    pub technology: String,
    pub state: String,
    pub signal: ModemSignal,
    pub sim: Option<SimInfo>,
    pub bearer: Option<ModemBearer>,
    pub registered: bool,
}

/// Radio signal quality of a modem.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct ModemSignal {
    pub quality: u8,
    pub rssi: Option<i32>,
    pub rsrp: Option<i32>,
    pub rsrq: Option<i32>,
    pub sinr: Option<i32>,
}

/// A SIM card inserted in a modem.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SimInfo {
    pub iccid: Option<String>,
    pub imsi: Option<String>,
    pub operator: Option<String>,
    pub mcc: Option<String>,
    pub mnc: Option<String>,
    pub number: Option<String>,
    pub state: String,
}

/// An active data bearer (PDP context) on a modem.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct ModemBearer {
    pub apn: Option<String>,
    pub ipv4: Option<String>,
    pub ipv6: Option<String>,
}

File: src/lib/core/types/modem.ts
import type { AccessTechnology, ModemState, SimState } from "./common";

export interface ModemSignal {
  quality: number;
  rssi: number | null;
  rsrp: number | null;
  rsrq: number | null;
  sinr: number | null;
}

export interface SimInfo {
  iccid: string | null;
  imsi: string | null;
  operator: string | null;
  mcc: string | null;
  mnc: string | null;
  number: string | null;
  state: SimState;
}

export interface ModemBearer {
  apn: string | null;
  ipv4: string | null;
  ipv6: string | null;
}

export interface ModemInfo {
  id: string;
  manufacturer: string;
  model: string;
  device: string;
  imei: string | null;
  operator: string | null;
  technology: AccessTechnology;
  state: ModemState;
  signal: ModemSignal;
  sim: SimInfo | null;
  bearer: ModemBearer | null;
  registered: boolean;
}

export interface ModemListResult {
  modems: ModemInfo[];
}
